src/lib/badge/badge.ts

```diff
@@ -0,0 +1,180 @@
+  'medium' |
+  'large';
+
+/** Directive to display a text badge. */
+@Directive({
+  selector: '[matBadge]',
+  inputs: ['matBadgeColor', 'matBadgeOverlap', 'matBadgePosition'],
+  host: {
+    'class': 'mat-badge',
+    '[class.mat-badge-overlap]': '_overlap',
+    '[class.mat-badge-above]': '_isAbove',
+    '[class.mat-badge-below]': '!_isAbove',
+    '[class.mat-badge-before]': '!_isAfter',
+    '[class.mat-badge-after]': '_isAfter',
+    '[class.mat-badge-small]': 'matBadgeSize === "small"',
+    '[class.mat-badge-medium]': 'matBadgeSize === "medium"',
+    '[class.mat-badge-large]': 'matBadgeSize === "large"',
+    '[class.mat-badge-hidden]': 'matBadgeHidden',
+  },
+})
+export class MatBadge {
+
+  /** The color of the badge. Can be `primary`, `accent`, or `warn`. */
+  @Input('matBadgeColor')
+  get color(): ThemePalette { return this._color; }
+  set color(value: ThemePalette) {
+    this._color = value;
+    this._setColor(value);
+  }
+  private _color: ThemePalette = 'primary';
+
+  /** Whether the badge should overlap its contents or not */
+  @Input('matBadgeOverlap')
+  get overlap(): boolean { return this._overlap; }
+  set overlap(val: boolean) {
+    this._overlap = coerceBooleanProperty(val);
+  }
+  private _overlap: boolean = true;
+
+  /**
+   * Position the badge should reside.
+   * Accepts any combination of 'above'|'below' and 'before'|'after'
+   */
+  @Input('matBadgePosition')
+  get position(): PositionTypes { return this._position; }
+  set position(val: PositionTypes) {
+    this._position = val;
+    this._isAbove = val.indexOf('below') === -1;
+    this._isAfter = val.indexOf('before') === -1;
+  }
+  private _position: PositionTypes = 'above after';
+
+  /** The content for the badge */
+  @Input('matBadge')
+  get content(): string { return this._content; }
+  set content(val: string) {
+    this._content = val;
+    this._updateTextContent();
+  }
+  private _content: string;
+
+  /** Message used to describe the decorated element via aria-describedby */
+  @Input('matBadgeDescription')
+  get description(): string { return this._description; }
+  set description(val: string) {
+    this._updateHostAriaDescription(val, this._description);
+    this._description = val;
+  }
+  private _description: string;
+
+  /** Size of the badge. Can be 'small' , 'medium', or 'large'. */
+  @Input() matBadgeSize: BadgeSize = 'medium';
+
+  /** Whether the badge is hidden. */
+  @Input()
+  set matBadgeHidden(val: boolean) {
+    this._hidden = coerceBooleanProperty(val);
+  }
+  get matBadgeHidden(): boolean {
+    return this._hidden;
+  }
+  private _hidden: boolean;
+
+  /** Unique id for the badge */
+  _id: number = nextId++;
+
+  /** Whether the badge is above the host or not */
+  _isAbove: boolean = true;
+
+  /** Whether the badge is after the host or not */
+  _isAfter: boolean = true;
+
+  private _badgeElement: HTMLElement;
+
+  constructor(
+      @Optional() @Inject(DOCUMENT) private _document: any,
```

Comment:
Instead of marking this as a class member w/ `private` here, we've been preferring to define the class member separately so that we can use the correct type (`Document`) throughout the rest of the code instead of `any`.

Reply:
Can you elaborate? Every other instance I've found in the code does it this way.

Reply:
```ts
class Example {
  private _document: Document;

  constructor(@Optional() @Inject(DOCUMENT) document: any) {
    this._document = document;
  }
}
```
Everywhere else was recently updated to do this so that referencing `_document` throughout the rest of the class is type-safe.